org.opent2t.sample.lamp.superpopular/com.hue.bulb/js/colour.js

```diff
@@ -0,0 +1,359 @@
+"use strict";
+
+const ChangeTolerance = 0.0001;
+const MaxHue = 360.0;
+const MaxColour = 255;
+
+/*
+ * Method to see if the given XY value is within the reach of the lamps.
+ *
+ * @param xyColour is a two-element object representing a point containing the X,Y value
+ * @param colourGamut   the color gamut of the light device.
+ * @return true if within reach, false otherwise.
+ */
+function isPointInLampsReach(xyColour, colourGamut){
+
+    var v1 = [colourGamut[1].x - colourGamut[0].x, colourGamut[1].y - colourGamut[0].y];
+    var v2 = [colourGamut[2].x - colourGamut[0].x, colourGamut[2].y - colourGamut[0].y];
+    var q = [xyColour.x - colourGamut[0].x, xyColour.y - colourGamut[0].y];
+
+    var s = product(q, v2) / product(v1, v2);
+    var t = product(v1, q) / product(v1, v2);
+
+    return ((s >= 0) && (t >= 0) && (s + t <= 1)) ? true : false;
+}
+
+/*
+ * Find the closest point on a line.
+ * This point will be within reach of the lamp.
+ *
+ * @param ptA the point where the line starts
+ * @param ptB the point where the line ends
+ * @param ptP the point which is close to a line.
+ * @return the point which is on the line.
+ */
+function getClosestPointToPoints(ptA, ptB, ptP){
+
+    var AP = {x: (ptP.x - ptA.x), y: (ptP.y - ptA.y)};
```

Comment:
Since those helper functions were derived from math equations, I am not sure how should I made them more descriptive. I added some more comments as well as references to make the code more readable. 

---
In reply to: [94657972](https://github.com/openT2T/translators/pull/143#discussion_r94657972) [](ancestors = 94657972,94657833,94528032,94471080)